### PR TITLE
feat: update tracers to use built in metrics

### PR DESCRIPTION
**Pull request by mutianf**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigtable/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+  private String cluster = "undefined";
+
+  // gfe stats
+  private AtomicLong gfeMissingHeaders = new AtomicLong(0);
+
+  BuiltinMetricsTracer(
+      ApiTracerFactory.OperationType operationType,
+      SpanName spanName,
+      Map<String, String> attributes,
+      StatsWrapper builtinMetricsWrapper,
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:64]: please be consistent between stats and metrics

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracerStreamingCallable.java [hunk 2]
```diff
@@ -26,26 +26,28 @@
...
  */
 @InternalApi
-public class HeaderTracerStreamingCallable<RequestT, ResponseT>
+public class BigtableTracerStreamingCallable<RequestT, ResponseT>
     extends ServerStreamingCallable<RequestT, ResponseT> {
 
   private final ServerStreamingCallable<RequestT, ResponseT> innerCallable;
 
-  public HeaderTracerStreamingCallable(
+  public BigtableTracerStreamingCallable(
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracerStreamingCallable.java:50]: any chance you can split this migration out into a separate PR? it creates a lot of noise

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracerStreamingCallable.java [hunk 3]
```diff
@@ -97,6 +100,16 @@ public void onError(Throwable t) {
...
+      //      try {
+      //        byte[] trailers =
+      //            responseMetadata
+      //                .getTrailingMetadata()
+      //                .get(Metadata.Key.of(Util.TRAILER_KEY, Metadata.BINARY_BYTE_MARSHALLER));
+      //        ResponseParams decodedTrailers = ResponseParams.parseFrom(trailers);
+      //        tracer.setLocations(decodedTrailers.getZoneId(), decodedTrailers.getClusterId());
+      //      } catch (NullPointerException | InvalidProtocolBufferException e) {
+      //      }
+
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracerStreamingCallable.java:112]: ?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracerStreamingCallable.java [hunk 4]
```diff
@@ -105,7 +118,42 @@ public void onComplete() {
...
       tracer.recordGfeMetadata(latency, null);
+      //      try {
+      //        byte[] trailers =
+      // responseMetadata.getTrailingMetadata().get(Metadata.Key.of(Util.TRAILER_KEY,
+      // Metadata.BINARY_BYTE_MARSHALLER));
+      //        ResponseParams decodedTrailers = ResponseParams.parseFrom(trailers);
+      //        tracer.setLocations(decodedTrailers.getZoneId(), decodedTrailers.getClusterId());
+      //      } catch (NullPointerException | InvalidProtocolBufferException e) {
+      //      }
+
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracerStreamingCallable.java:129]: ?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracerUnaryCallable.java [hunk 5]
```diff
@@ -82,13 +82,32 @@ public void onFailure(Throwable throwable) {
...
+      //      try {
+      //        byte[] trailers =
+      //                responseMetadata
+      //                        .getTrailingMetadata()
+      //                        .get(Metadata.Key.of(Util.TRAILER_KEY,
+      // Metadata.BINARY_BYTE_MARSHALLER));
+      //        ResponseParams decodedTrailers = ResponseParams.parseFrom(trailers);
+      //        tracer.setLocations(decodedTrailers.getZoneId(), decodedTrailers.getClusterId());
+      //      } catch (NullPointerException | InvalidProtocolBufferException e) {
+      //      }
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracerUnaryCallable.java:94]: ?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracerUnaryCallable.java [hunk 6]
```diff
@@ -82,13 +82,32 @@ public void onFailure(Throwable throwable) {
...
       tracer.recordGfeMetadata(latency, null);
+      //      try {
+      //        byte[] trailers =
+      //            responseMetadata
+      //                .getTrailingMetadata()
+      //                .get(Metadata.Key.of(Util.TRAILER_KEY, Metadata.BINARY_BYTE_MARSHALLER));
+      //        ResponseParams decodedTrailers = ResponseParams.parseFrom(trailers);
+      //        tracer.setLocations(decodedTrailers.getZoneId(), decodedTrailers.getClusterId());
+      //      } catch (NullPointerException | InvalidProtocolBufferException e) {
+      //      }
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracerUnaryCallable.java:110]: ?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 7]
```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright 2021 Google LLC
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:2]: 2022

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+import com.google.common.base.Stopwatch;
+import java.util.Map;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+import org.threeten.bp.Duration;
+
+public class BuiltinMetricsTracer extends BigtableTracer {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:31]: docs & visibility

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+
+  private BuiltinMetricsRecorder recorder;
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:33]: final

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerFactory.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2021 Google LLC
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerFactory.java:2]: here and everywhere else 2022

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/Util.java [hunk 11]
```diff
@@ -68,14 +78,14 @@ static TagValue extractStatus(@Nullable Throwable error) {
...
-    return TagValue.create(statusString);
+    return statusString;
   }
 
   /**
    * Await the result of the future and convert it into a value that can be used as an OpenCensus
    * tag value.
    */
-  static TagValue extractStatus(Future<?> future) {
+  static TagValue extractStatusAsync(Future<?> future) {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/Util.java:88]: but this isnt async :)

File: pom.xml [hunk 12]
```diff
@@ -212,6 +212,15 @@
...
     <build>
+        <pluginManagement>
+            <plugins>
+                <plugin>
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-shade-plugin</artifactId>
+                    <version>3.2.4</version>
+                </plugin>
+            </plugins>
+        </pluginManagement>
```
Comment (igorbernstein2) [inline pom.xml:223]: this should probably move to the previous PR

File: google-cloud-bigtable-bom/pom.xml [hunk 13]
```diff
@@ -94,6 +94,11 @@
           <artifactId>proto-google-cloud-bigtable-v2</artifactId>
           <version>2.8.1-SNAPSHOT</version><!-- {x-version-update:proto-google-cloud-bigtable-v2:current} -->
         </dependency>
+          <dependency>
+              <groupId>com.google.cloud</groupId>
+              <artifactId>google-cloud-bigtable-stats</artifactId>
+              <version>2.8.1-SNAPSHOT</version><!-- {x-version-update:proto-google-cloud-bigtable-v2:current} -->
+          </dependency>
```
Comment (igorbernstein2) [inline google-cloud-bigtable-bom/pom.xml:101]: move this to previous PR

File: google-cloud-bigtable-stats/pom.xml [hunk 14]
```diff
@@ -79,6 +79,7 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-shade-plugin</artifactId>
+                <version>3.2.4</version>
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/pom.xml:82]: please move to prev pr

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracer.java [hunk 15]
```diff
@@ -57,4 +62,9 @@ public void recordGfeMetadata(@Nullable Long latency, @Nullable Throwable throwa
   public void batchRequestThrottled(long throttledTimeMs) {
     // noop
   }
+
+  /** Set the Bigtable zone and cluster so metrics can be tagged with location information. */
+  public void setLocations(String zone, String cluster) {
+    // noop
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BigtableTracer.java:68]: add a comment that this is used by builtin metrics

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+
+  BuiltinMetricsTracer(
+      ApiTracerFactory.OperationType operationType,
+      SpanName spanName,
+      Map<String, String> attributes,
+      StatsWrapper statsWrapper,
+      StatsRecorderWrapper recorder) {
+    this.operationType = operationType;
+    this.spanName = spanName;
+    if (recorder != null) {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:72]: when is recorder null?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+  private final Stopwatch firstResponsePerOpTimer = Stopwatch.createStarted();
+
+  // Attempt level metrics
+  private int attemptCount = 0;
+  private Stopwatch attemptTimer;
+  private volatile int attempt = 0;
+
+  // Total application latency
+  private final Stopwatch applicationLatencyTimer = Stopwatch.createUnstarted();
+  private final AtomicLong totalApplicationLatency = new AtomicLong(0);
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:54]: PLease add a comment why this needs to be atomic

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+
+  @Override
+  public void attemptStarted(Object request, int attemptNumber) {
+    this.attempt = attemptNumber;
+    attemptCount++;
+    attemptTimer = Stopwatch.createStarted();
+    if (request != null) {
+      this.tableId = Util.extractTableId(request);
+    }
+    if (applicationLatencyTimer.isRunning()) {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:115]: This is inconsistent by using an atomic for applicationLatency I assume that it can be written and read by different threads which is true (gRPC thread for attemptFailed and user thread for request()).
But this would also be true for the applicationLatencyTimer? should there be something to ensure consistency?

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java [hunk 19]
```diff
@@ -0,0 +1,341 @@
...
+                    headers.put(
+                        Metadata.Key.of("server-timing", Metadata.ASCII_STRING_MARSHALLER),
+                        String.format("gfet4t7; dur=%d", FAKE_SERVER_TIMING));
+                    super.sendHeaders(headers);
+                  }
+
+                  @Override
+                  public void close(Status status, Metadata trailers) {
+                    super.close(status, trailers);
+                  }
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:163]: is this override needed? Won't SimpleForwardingServerCall handle it for you?

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java [hunk 20]
```diff
@@ -0,0 +1,341 @@
...
+    stub = new EnhancedBigtableStub(stubSettings, ClientContext.create(stubSettings));
+
+    serviceHelper.start();
+
+    longValue = ArgumentCaptor.forClass(long.class);
+    intValue = ArgumentCaptor.forClass(int.class);
+    status = ArgumentCaptor.forClass(String.class);
+    tableId = ArgumentCaptor.forClass(String.class);
+    zone = ArgumentCaptor.forClass(String.class);
+    cluster = ArgumentCaptor.forClass(String.class);
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:199]: I think its cleaner to annotate the member variables with `@Captor` and use the mockito rule to construct em

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java [hunk 21]
```diff
@@ -0,0 +1,341 @@
...
+                    super.close(status, trailers);
+                  }
+                },
+                metadata);
+          }
+        };
+
+    serviceHelper = new FakeServiceHelper(trailersInterceptor, mockService);
+
+    statsRecorderWrapper = Mockito.mock(StatsRecorderWrapper.class);
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:171]: I would use the mockito rule and annotate the member variable with `@Mock`

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java [hunk 22]
```diff
@@ -0,0 +1,341 @@
...
+  @Test
+  public void testMutateRowAttempts() {
+    stub.mutateRowCallable()
+        .call(RowMutation.create(TABLE_ID, "random-row").setCell("cf", "q", "value"));
+
+    verify(statsRecorderWrapper, times(4))
+        .record(status.capture(), tableId.capture(), zone.capture(), cluster.capture());
+    assertThat(zone.getAllValues().get(0)).isEqualTo(UNDEFINED);
+    assertThat(zone.getAllValues().get(1)).isEqualTo(UNDEFINED);
+    assertThat(zone.getAllValues().get(2)).isEqualTo(UNDEFINED);
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:300]: here and everywhere else:
`assertThat(zone.getAllValues().containsExactly(UNDEFINED, UNDEFINED, UNDEFINED).inOrder()`

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java [hunk 23]
```diff
@@ -0,0 +1,341 @@
...
+      responseObserver.onCompleted();
+    }
+
+    @Override
+    public void mutateRow(
+        MutateRowRequest request, StreamObserver<MutateRowResponse> responseObserver) {
+      if (serverRetryDelayStopwatch.isRunning()) {
+        serverTotalRetryDelay.addAndGet(serverRetryDelayStopwatch.elapsed(TimeUnit.MILLISECONDS));
+        serverRetryDelayStopwatch.reset();
+      }
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:330]: I think you might cache coherences issues here: the thread that create serverRetryDelayStopwatch can be different from the current one and the values might not have been flushed to memory

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 24]
```diff
@@ -0,0 +1,251 @@
...
+
+  // Attempt level metrics
+  private int attemptCount = 0;
+  private Stopwatch attemptTimer;
+  private volatile int attempt = 0;
+
+  // Total application latency
+  // Total application latency needs to be atomic because it's accessed from different threads. E.g.
+  // request() from
+  // user thread and attempt failed from grpc thread.
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:55]: please fix formatting

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 25]
```diff
@@ -0,0 +1,251 @@
...
+      StatsWrapper statsWrapper,
+      @Nullable StatsRecorderWrapper statsRecorderWrapper) {
+    this.operationType = operationType;
+    this.spanName = spanName;
+    if (statsRecorderWrapper != null) {
+      // A workaround for test to pass in a mock StatsRecorderWrapper
+      this.recorder = statsRecorderWrapper;
+    } else {
+      this.recorder = new StatsRecorderWrapper(operationType, spanName, attributes, statsWrapper);
+    }
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:83]: I think you should add a factory method and avoid if statements for test behavior. Also, if add the factory method to statsWrapper like i mentioned in the other PR, it will look something liek this:

```java
static BuiltinMetricsTracer create(OperationType, SpanName, ...) {
   return new BuiltinMetricsTracer(opType, spanName, attribs, statsWrapper.createRecorder(...));
}
// VisibleForTesting
BuiltinMetricsTracer(OperationType, SpanName, ..., StatsRecorderWrapper recorder) {
}
```

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+  }
+
+  @Override
+  public void attemptPermanentFailure(Throwable error) {
+    super.attemptPermanentFailure(error);
+  }
+
+  @Override
+  public void lroStartFailed(Throwable error) {
+    super.lroStartFailed(error);
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:158]: here and everywhere else, why do you need to the super invocations?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerFactory.java [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+  private final StatsRecorderWrapper statsRecorderWrapper;
+
+  public static BuiltinMetricsTracerFactory create(
+      StatsWrapper statsWrapper, ImmutableMap<String, String> statsAttributes) {
+    return new BuiltinMetricsTracerFactory(statsWrapper, statsAttributes, null);
+  }
+
+  // A workaround for test to pass in a mock StatsRecorderWrapper
+  @VisibleForTesting
+  static BuiltinMetricsTracerFactory createWithRecorder(
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerFactory.java:46]: is this necessary? I think you can just have your tests use the constructor directly. In general I like the pattern of having a public factory method for production consumers and a package protected ctor for tests

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/HeaderTracerStreamingCallable.java [hunk 28]
```diff
@@ -82,7 +82,8 @@ public void call(
 
     @Override
     public void onStart(final StreamController controller) {
-      outerObserver.onStart(controller);
+      final StreamController tracedController = new TracedStreamController(controller, tracer);
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/HeaderTracerStreamingCallable.java:85]: final?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java [hunk 29]
```diff
@@ -0,0 +1,251 @@
...
+
+  @Override
+  public void responseReceived() {
+    if (applicationLatencyTimerIsRunning.compareAndSet(false, true)) {
+      applicationLatencyTimer.start();
+    }
+    if (firstResponsePerOpTimer.isRunning()) {
+      firstResponsePerOpTimer.stop();
+    }
+  }
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java:182]: I think your application latency calculations are a bit off. What happens when the caller disables auto flow control and request(10)?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/MetricsTracer.java [hunk 30]
```diff
@@ -248,4 +253,14 @@ private TagContextBuilder newTagCtxBuilder() {
...
+
+  @Override
+  public void setLocations(String zone, String cluster) {
+    // noop
+  }
+
+  @Override
+  public void onRequest() {
+    // noop
+  }
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/MetricsTracer.java:265]: why do you mean empty overrides?

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java [hunk 31]
```diff
@@ -0,0 +1,310 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+import org.threeten.bp.Duration;
+
+public class BuiltinMetricsTracerTest {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:69]: `RunWith(Junit4.class)`

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java [hunk 32]
```diff
@@ -0,0 +1,401 @@
...
+  private BigtableGrpc.BigtableImplBase mockService;
+  private Server server;
+
+  private EnhancedBigtableStub stub;
+
+  @Mock private BuiltinMetricsTracerFactory mockFactory;
+  @Mock private StatsRecorderWrapper statsRecorderWrapper;
+
+  @Captor private ArgumentCaptor<Long> longValue;
+  @Captor private ArgumentCaptor<Integer> intValue;
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:103]: Please use better names for these. Perhaps latency & retryCount?

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java [hunk 33]
```diff
@@ -0,0 +1,401 @@
...
+        .thenAnswer(
+            (Answer<BuiltinMetricsTracer>)
+                invocationOnMock ->
+                    new BuiltinMetricsTracer(
+                        OperationType.ServerStreaming,
+                        SpanName.of("Bigtable", "ReadRows"),
+                        statsRecorderWrapper));
+
+    final long applicationLatency = 200;
+    final AtomicInteger counter = new AtomicInteger(0);
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:264]: why is this atomic?

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java [hunk 34]
```diff
@@ -0,0 +1,401 @@
...
+      Thread.sleep(applicationLatency);
+      rows.next();
+    }
+
+    verify(statsRecorderWrapper).putApplicationLatencies(longValue.capture());
+
+    // For manual flow control, the last application latency shouldn't count, because at that point
+    // the server already sent back all the responses.
+    assertThat(counter.get()).isGreaterThan(1);
+    assertThat(longValue.getValue())
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:278]: Also, to make it easier to read, please assign longValue.get() to totalApplicationLatency and use that variable

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java [hunk 35]
```diff
@@ -0,0 +1,401 @@
...
+                        OperationType.ServerStreaming,
+                        SpanName.of("Bigtable", "ReadRows"),
+                        statsRecorderWrapper));
+
+    stub.mutateRowCallable()
+        .call(RowMutation.create(TABLE_ID, "random-row").setCell("cf", "q", "value"));
+
+    verify(statsRecorderWrapper).putRetryCount(intValue.capture());
+
+    assertThat(intValue.getValue()).isEqualTo(3);
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:300]: Please create a named constant for retry counts...its hard to understand why 3 is correct here.
Also is it intended that retry count == attempt count? ie does the first attempt count as a retry?

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java (hunk 33 above)
Comment (mutianf) [reply google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java:264]: Hmm, because it's used in response observer so it needs to be final, and I also need to update the value, so i need to  AtomicInteger here?

**Review by igorbernstein2 — APPROVED** [review]
lgtm
